crates/llm-ls/src/retrieval.rs

```diff
@@ -459,31 +480,63 @@ impl SnippetRetriever {
             let window = lines[start_line..end_line].to_vec();
             let snippet = window.join("\n");
             if snippet.is_empty() {
-                continue;
+                    debug!("snippet {file_url}[{start_line}, {end_line}] empty");
+                    continue;
             }
+            snippets.push(Snippet{ file_url: file_url.clone().into(), code: snippet, start_line, end_line });
+        }
+        {
+            let nb_snippets = snippets.len();
+            let steps = self.window_step;
+            debug!("Found {nb_snippets} snippets in {file_url}: {start}, {end}, {steps}");
+        }
 
-            let mut encoding = self.tokenizer.encode(snippet.clone(), true)?;
-            encoding.truncate(512, 1, TruncationDirection::Right);
-            let result = self.generate_embedding(encoding, self.model.clone()).await;
-            let embedding = match result {
-                Ok(e) => e,
+        // Group by length to reduce padding effect
+        snippets.sort_by(|first, second| first.code.len().cmp(&second.code.len()));
+        // TODO improvements to compute an efficient batch size:
+        //  - batch size should be relative to the cumulative size of all elements in the batch,
+        // Set embedding_batch_size to 8 if device is GPU, use match
+        let embedding_batch_size = match self.model.device {
+            Device::Cpu => 1,
+            _ => 8,
+        };
+        for batch in snippets.chunks(embedding_batch_size) {
+            let batch_code = batch
+                .iter()
+                .map(|snippet| snippet.code.clone())
+                .collect();
+            let encodings = self.tokenizer
+                .encode_batch(batch_code, true)?
+                .iter_mut()
+                .map(|encoding| {
+                    encoding.truncate(512, 1, TruncationDirection::Right);
+                    encoding.clone()
+                })
+                .collect();
+            let results = match self.generate_embedding(encodings, self.model.clone()).await {
+                Ok(result) => result,
                 Err(err) => {
-                    error!(
-                        "error generating embedding for {file_url}[{start_line}, {end_line}]: {err}",
-                    );
-                    continue;
+                    debug!("Unable to generate embeddings because of {:?}", err);
```

Comment:
```suggestion
                    debug!("generate embeddings error: {err}");
```

I'm pretty sure the error type implements `Display`